cache/lru_cache.cc

```diff
@@ -398,11 +401,7 @@ void LRUCacheShard::Promote(LRUHandle* e) {
   if (e->value) {
     Cache::Handle* handle = reinterpret_cast<Cache::Handle*>(e);
     Status s = InsertItem(e, &handle, /*free_handle_on_fail=*/false);
-    if (s.ok()) {
-      // InsertItem would have taken a reference on the item, so decrement it
-      // here as we expect the caller to already hold a reference
-      e->Unref();
-    } else {
+    if (!s.ok()) {
```

Comment:
Before calling InsertItem(), the ref count on e is 1. When thread 1 called Lookup(), it would have allocated e and taken a reference. Later when thread 1 calls Promote(), InsertItem() is called which will insert e in the handle table and increment the ref count to 2. To compensate for that, thread 1 then calls Unref() after InsertItem() returns, so ref count becomes 1 again.

The race happens if thread 2 calls Lookup(), gets e, and calls Ref() at the same time thread 1 is calling Unref() in Promote(). Since refs_ is not atomic, the value could be inconsistent.

Reply:
Got it. Thanks for the explanation.